x-pack/plugins/secops/public/components/range_date_picker/index.tsx

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { EuiDatePickerRange, EuiFlexGroup, EuiFlexItem, EuiFormControlLayout } from '@elastic/eui';
+import { get, getOr } from 'lodash/fp';
+import moment, { Moment } from 'moment';
+import React from 'react';
+import { connect } from 'react-redux';
+import { inputsActions, inputsModel, State } from '../../store';
+import { GlobalDateButton } from './global_date_button';
+import { QuickSelectPopover } from './quick_select_popover';
+import { UpdateButton } from './update_button';
+
+export type DateType = 'relative' | 'absolute';
+
+interface RangeDatePickerStateRedux {
+  from: number;
+  to: number;
+  isTimerOn: boolean;
+  duration: number;
+  isLoading: boolean;
+  refetch: inputsModel.Refetch[];
+}
+
+interface RangeDatePickerDispatchProps {
+  setAbsoluteRangeDatePicker: (params: { id: string; from: number; to: number }) => void;
+  setRelativeRangeDatePicker: (
+    params: { id: string; option: string; from: number; to: number }
+  ) => void;
+  startAutoReload: (params: { id: string }) => void;
+  stopAutoReload: (params: { id: string }) => void;
+  setDuration: (params: { id: string; duration: number }) => void;
+}
+interface OwnProps {
+  id: string;
+  disabled?: boolean;
+}
+
+type RangeDatePickerProps = OwnProps & RangeDatePickerDispatchProps & RangeDatePickerStateRedux;
+
+export interface RecentlyUsedI {
+  kind: string;
+  text: string | number[];
+}
+
+interface RangeDatePickerState {
+  recentlyUsed: RecentlyUsedI[];
+}
+
+class RangeDatePickerComponents extends React.PureComponent<
+  RangeDatePickerProps,
+  RangeDatePickerState
+> {
+  public readonly state = {
+    recentlyUsed: [] as RecentlyUsedI[],
```

Comment:
Here is an `as` that is where TypeScript types kind of break down or at least allow things that should be impossible, possible.

If this were changed into:

```ts
  public readonly state: { recentlyUsed: RecentlyUsedI } = {
    recentlyUsed: [],
  };
```

without using the `as` cast, then TypeScript rightfully throws an error that it cannot use an empty array since that type:

```ts
export interface RecentlyUsedI {
   kind: string;
   text: string | number[];
 }
```

states that those values are _not_ optional. At some point someone might end up using this default state and blow up during runtime.

I suggest either making those elements optional on the type if they truly can be optionally set, or set them with default values when the state is initialized if their initial states are to be set.

Reply:
I will change to make it more clear